## [作業重點]
目前你應該已經要很清楚資料集中，資料的型態是什麼樣子囉！包含特徵 (features) 與標籤 (labels)。因此要記得未來不管什麼專案，必須要把資料清理成相同的格式，才能送進模型訓練。
今天的作業開始踏入決策樹這個非常重要的模型，請務必確保你理解模型中每個超參數的意思，並試著調整看看，對最終預測結果的影響為何

# 作業

## 1. 試著調整 DecisionTreeClassifier(...) 中的參數，並觀察是否會改變結果？

In [1]:
from sklearn import datasets, metrics

# 如果是分類問題，請使用 DecisionTreeClassifier，若為回歸問題，請使用 DecisionTreeRegressor
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.model_selection import train_test_split

In [2]:
# 讀取鳶尾花資料集
iris = datasets.load_iris()
# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.25, random_state=4)

In [3]:
# 建立模型
clf = DecisionTreeClassifier()
# 訓練模型
clf.fit(x_train, y_train)
# 預測測試集
y_pred = clf.predict(x_test)

In [4]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9736842105263158


In [5]:
clf = DecisionTreeClassifier(criterion='entropy', max_depth=10, min_samples_leaf=1, min_samples_split=2)
clf.fit(x_train, y_train)
y_pred = clf.predict(x_test)
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9736842105263158


> ANS:　加上參數後沒有改變。criterion='entropy', max_depth=10, min_samples_leaf=1, min_samples_split=2

## 2. 改用其他資料集 (boston, wine)，並與回歸模型的結果進行比較

In [6]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets, linear_model
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error, r2_score

In [7]:
wine = datasets.load_wine()
print("feature:", wine.feature_names)
print("shape:", wine.data.shape, wine.target.shape)

feature: ['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
shape: (178, 13) (178,)


In [8]:
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.3, random_state=4)

In [9]:
wine_clf = DecisionTreeClassifier(criterion='entropy', max_depth=10, min_samples_leaf=1, min_samples_split=2)
wine_clf.fit(x_train, y_train)
y_pred = wine_clf.predict(x_test)
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print("Cross-validation: ", cross_val_score(wine_clf, x_train, y_train, cv=5).mean())
print("Mean squared error: %.2f" % mean_squared_error(y_test, y_pred))

Accuracy:  0.9074074074074074
Cross-validation:  0.9111851851851853
Mean squared error: 0.15


In [10]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [11]:
print("Feature importance: ", wine_clf.feature_importances_)

Feature importance:  [0.         0.         0.         0.02126655 0.09577741 0.
 0.0333229  0.         0.         0.10261596 0.         0.41837057
 0.32864662]


### Logistic Regression

In [12]:
regr = linear_model.LogisticRegression()
regr.fit(x_train, y_train)
y_pred = regr.predict(x_test)
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)
print(regr.coef_)
print("Cross-validation: ", cross_val_score(regr, x_train, y_train, cv=5).mean())
print("Mean squared error: %.2f" % mean_squared_error(y_test, y_pred))

Accuracy:  0.9444444444444444
[[-3.35652407e-01  8.72426703e-01  6.44807906e-01 -7.93275076e-01
  -4.64265711e-02  4.06656913e-01  1.00616444e+00 -7.54314830e-02
  -2.07146703e-01  2.67096434e-02 -1.56752009e-01  8.17113637e-01
   1.94064438e-02]
 [ 8.23877593e-01 -1.14169684e+00 -2.83474220e-01  3.60915096e-01
  -4.67507982e-03 -3.31329565e-02  4.83545668e-01  3.97082413e-01
   7.44567099e-01 -2.18398714e+00  7.45228311e-01  4.06364841e-01
  -1.27242260e-02]
 [-4.14834564e-01  5.45124220e-01 -1.27288118e-01  1.01220680e-01
   4.37310660e-02 -5.76415265e-01 -1.44951324e+00 -6.42241624e-02
  -7.06140809e-01  1.15191616e+00 -4.36353167e-01 -1.26993332e+00
  -1.90141418e-03]]
Cross-validation:  0.975925925925926
Mean squared error: 0.06


/home/ziv/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/home/ziv/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
/home/ziv/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/home/ziv/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
/home/ziv/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: 

### Linear Regression w/ LASSO

In [13]:
lasso = linear_model.Lasso(alpha=1.0)
lasso.fit(x_train, y_train)
y_pred = lasso.predict(x_test)
#acc = metrics.accuracy_score(y_test, y_pred)
#print("Acuuracy: ", acc)
print(lasso.coef_)
print("Cross-validation: ", cross_val_score(lasso, x_train, y_train, cv=5).mean())
print("Mean squared error: %.2f" % mean_squared_error(y_test, y_pred))

[ 0.          0.          0.          0.          0.         -0.
 -0.          0.         -0.          0.         -0.         -0.
 -0.00161519]
Cross-validation:  0.39047558279793987
Mean squared error: 0.41


### Linear Regression w/ Ridge

In [14]:
ridge = linear_model.Ridge(alpha=1.0)
ridge.fit(x_train, y_train)
y_pred = ridge.predict(x_test)
#acc = metrics.accuracy_score(y_test, y_pred)
#print("Acuuracy: ", acc)
print(ridge.coef_)
print("Cross-validation: ", cross_val_score(ridge, x_train, y_train, cv=5).mean())
print("Mean squared error: %.2f" % mean_squared_error(y_test, y_pred))

[-0.09688913  0.0134691  -0.17227376  0.04501149  0.00133898  0.10829142
 -0.3000121  -0.05292874  0.04386404  0.07697862 -0.10493811 -0.29315367
 -0.00083382]
Cross-validation:  0.867384027972468
Mean squared error: 0.07


|  | DecisionTree | Logistic | Linear w/ LASSO | Linear w/ Ridge |
|---|---|---|---|---|
| Accuracy         | 0.92 | 0.94 | N/A  | N/A  |
| Cross-validation | 0.93 | 0.86 | 0.39 | 0.86 |
| MSE              | 0.04 | 0.07 | 0.41 | 0.07 |